controller/src/main/java/io/pravega/controller/store/stream/ZookeeperBucketStore.java

```diff
@@ -0,0 +1,261 @@
+                        pathChildrenCache.start(PathChildrenCache.StartMode.BUILD_INITIAL_CACHE);
+                    } catch (Exception e) {
+                        log.error("Starting ownership listener for service {} threw exception {}", serviceType, e);
+                        throw Lombok.sneakyThrow(e);
+                    }
+                    log.info("bucket ownership listener registered on bucket root {}", serviceType);
+
+                    return pathChildrenCache;
+                });
+        }
+
+    @Override
+    public synchronized void unregisterBucketOwnershipListener(ServiceType serviceType) {
+        PathChildrenCache pathChildrenCache = bucketOwnershipCacheMap.remove(serviceType);
+        if (pathChildrenCache != null) {
+            try {
+                pathChildrenCache.clear();
+                pathChildrenCache.close();
+            } catch (IOException e) {
+                log.warn("unable to close listener for bucket ownership {}", e);
+            }
+        }
+    }
+
+    @Override
+    public synchronized void registerBucketChangeListener(ServiceType serviceType, int bucket, BucketChangeListener listener) {
+        Preconditions.checkNotNull(listener);
+
+        PathChildrenCacheListener bucketListener = (client, event) -> {
+            StreamImpl stream;
+            switch (event.getType()) {
+                case CHILD_ADDED:
+                    stream = getStreamFromPath(event.getData().getPath());
+                    listener.notify(new StreamNotification(stream.getScope(), stream.getStreamName(), NotificationType.StreamAdded));
+                    break;
+                case CHILD_REMOVED:
+                    stream = getStreamFromPath(event.getData().getPath());
+                    listener.notify(new StreamNotification(stream.getScope(), stream.getStreamName(), NotificationType.StreamRemoved));
+                    break;
+                case CHILD_UPDATED:
+                    stream = getStreamFromPath(event.getData().getPath());
+                    listener.notify(new StreamNotification(stream.getScope(), stream.getStreamName(), NotificationType.StreamUpdated));
+                    break;
+                case CONNECTION_LOST:
+                    listener.notify(new StreamNotification(null, null, BucketChangeListener.StreamNotification.NotificationType.ConnectivityError));
+                    break;
+                default:
+                    log.warn("Received unknown event {} on bucket", event.getType(), bucket);
+            }
+        };
+
+        String bucketPath = getBucketPath(serviceType.getName(), bucket);
+        
+        bucketCacheMap.computeIfAbsent(bucketPath, x -> {
+            PathChildrenCache pathChildrenCache = new PathChildrenCache(storeHelper.getClient(), bucketPath, true);
+            pathChildrenCache.getListenable().addListener(bucketListener);
+            try {
+                pathChildrenCache.start(PathChildrenCache.StartMode.NORMAL);
```

Comment:
This looks like it is performing IO under both the method lock and the ConcurrentHashMap's lock. This is not only bad in it's own right, if it is invoking the listener callback inside of here, it could deadlock.

Reply:
Did not intend to have "synchronized" in the method. That was a slip. 
However, i did miss the fact that we are doing IO under the lock in concurrent hash map. 
As we are thinking of refactoring to improve the scalability, this code should be fixed. 
Let me make those changes and i will update it here. Thanks.